syntax_and_semantics/annotations/user_defined_annotations.md

```diff
@@ -0,0 +1,174 @@
+
+**NOTE:** If multiple annotations of the same type are applied, the `annotation` method will return the _last_ one.
+
+The [@type](../macros.md#type-information) and [@def](../macros.md#method-information) variables can be used to get `TypeNode` or `Def` object to use the `annotation` method on.  However, it is also possible to get `TypeNode`/`Def` types using other methods on `TypeNode`.  For example `TypeNode.all_subclasses` or `TypeNode.methods`, respectively.
+
+The `TypeNode.instance_vars` can be used to get an array of instance variable `MetaVar` objects that would allow reading annotations defined on those instance variables.
+
+**NOTE:** `TypeNode.instance_vars` currently only works in the context of an instance/class method.
+
+```crystal
+annotation MyClass; end
+annotation MyMethod; end
+annotation MyIvar; end
+
+@[MyClass]
+class Foo
+  pp {{@type.annotation(MyClass).stringify}}
+  
+  @[MyIvar]
+  @num : Int32 = 1
+  
+  @[MyIvar]
+  property name : String = "jim"
+
+  def properties
+    {% for ivar in @type.instance_vars %}
+      pp {{ivar.annotation(MyIvar).stringify}}
+    {% end %}
+  end
+end
+
+@[MyMethod]
+def my_method
+  pp {{@def.annotation(MyMethod).stringify}}
+end
+
+Foo.new.properties
+my_method
+pp {{Foo.annotation(MyClass).stringify}}
+
+# Which would print
+"@[MyClass]"
+"@[MyIvar]"
+"@[MyIvar]"
+"@[MyMethod]"
+"@[MyClass]"
+```
+
+### Reading Multiple Annotations
+
+If there are multiple annotations of the same type applied to the same instance variable/method/type, the `annotations(type : TypeNode)` method can be used.  This will work on anything that `annotation(type : TypeNode)` would, but instead returns an `ArrayLiteral(Annotation)`.
+
+```crystal
+annotation MyAnnotation; end
+
+@[MyAnnotation("foo")]
+@[MyAnnotation(123)]
+@[MyAnnotation(123)]
+def annotation_read
+  {% for ann, idx in @def.annotations(MyAnnotation) %}
+    pp "Annotation {{idx}} = {{ann[0].id}}"
+  {% end %}
+end
+
+annotation_read
+
+# Which would print
+"Annotation 0 = foo"
+"Annotation 1 = 123"
+"Annotation 2 = 123"
+```
+
+## Fields
+
+Data can be stored within an annotation.
+
+```crystal
+annotation MyAnnotaion; end
+
+# The fields can either be a key/value pair
+@[MyAnnotation(key: "value", value: 123)]
+
+# Or positional
+@[MyAnnotation("foo", 123, false)]
+```
+
+### Key/value
+
+The values of annotation key/value pairs can be accessed at compile time via the [`[]`](https://crystal-lang.org/api/Crystal/Macros/Annotation.html#%5B%5D%28name%3ASymbolLiteral%7CStringLiteral%7CMacroId%29%3AASTNode-instance-method) method.
+
+```crystal
+annotation MyAnnotation; end
+
+@[MyAnnotation(default: 2)]
+def double(num : Number? = nil)
+  # The name can be a `String`, `Symbol`, or `MacroId`
+  2 * (num ? num : {{@def.annotation(MyAnnotation)[:default]}})
+end
```

Comment:
This exemple is soo weird, the end goal would have been simpler using a proper default value in the method definition!

I know I don't have any suggestion, but I think we should try to find a more interesting exemple here

Reply:
Yea....However it's kinda hard to come up with actual examples that annotations would be useful in, that is simple enough to show.  So, if you have any ideas let me know :)

Reply:
Probably just print the `2` value if you can't think of a "real" example.